telescope.yaml

```diff
@@ -0,0 +1,121 @@
+# This file creates replicas number of Pods, all served with one Kubernetes Service resource
+# each pod has 2 containers (telescope and redis) as has been specified in following Deployment resource
+
+# Change <DockerID> with the Docker Hub namespace (ID/username) in this file
+# To deploy project on Kubernetes we must have telescope image from `Dockerfile-deploy` file pushed on Docker Hub (instruction in `Dockerfile-deploy`).
+# We need `kubectl` installed, and for local simulation of the deployment, `minikube` installed and running.
+# deploy the app using following command
+# $ kubectl create -f telescope.yaml
+# to watch the pods as they gets created
+# $ kubectl get pods --watch
+# to see stdout of containers (telescope or redis) in one specific instance (pod)
+# $ kubctl logs <pod name as shown by above command> -c [telescope | redis]   ---> ex. $ kubectl logs telescope-7b567445fc-pn7cb -c telescope
+# finally access to the app which is running in Kubernetes cluster using this command
+# $ minikube service telescope
+# we can scale the app using the following command (example: from 3 replicas to 5)
+# $ kubectl scale --replicas=5 telescope
+# re-creating and replacing resources defined in a YAML file (ONLY IN MINIKUBE):
+# $ kubectl replace --force -f telescope.yaml   -----> Will cause service outage; Do not use in real deployment in Kubernetes
+
+# Service resource
+## telescope needs a service resource to be accessible to the outside the cluster
+## Service resource also guarantees the same IP address after each restart
+
+# version of Kubernetes API- using stable version
+apiVersion: v1
+# resource type and label
+kind: Service
+metadata:
+  name: telescope
+# specifications of resource
+spec:
+  selector:
+    name: telescope
+  ports:
+    # the port that cluster is listening (exposed)
+    - port: 3000
+      # the port in each pod that Service maps traffic on
+      targetPort: 3000
+      # the exposed port will be manipulated as nodePort. If we don't define it, it would be a random number greater than/equal 30000
+      # this is the port number we will see when we access to service by cluster's IP address
+      nodePort: 30000
+  # This resource needs to communicate with outside of the cluster
+  # and share traffic between instances of app
+  type: LoadBalancer
```

Comment:
Although this "works" for minikube, we should change this to NodePort, since we don't actually have a load balancer

Reply:
This is a LoadBalancer and I have it here by purpose. I actually commented about the reason right at the line above: `share traffic between instances of app`. We have 3 replicas of the app (line 58 of this file) running at the same time.

Reply:
I think there's a difference between having replicas and actually load balancing. If you want to use the load balancing service, you need to use ```minikube tunnel``` as well (I think).

https://minikube.sigs.k8s.io/docs/tasks/loadbalancer/
https://blog.codonomics.com/2019/02/loadbalancer-support-with-minikube-for-k8s.html